### PR TITLE
Add getter/setter of C# OrtEnv log level

### PR DESCRIPTION
**Pull request by yf711**
### Description
* Add getter/setter to access and update C# OrtEnv log level
* Add C API about updating ort env with custom log level to support the setter above (Following [pybind implementation](https://github.com/microsoft/onnxruntime/blob/952c99304a764328070ac6c07f510deec06dedf4/onnxruntime/python/onnxruntime_pybind_state.cc#L923-L924))
* Add test case to verify getter & setter

### Motivation and Context
* For C++/Python, the log level can be adjusted via OrtEnv, and this feature is missing in C# binding

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -3590,6 +3590,14 @@ struct OrtApi {
...
   void(ORT_API_CALL* MemoryInfoGetDeviceType)(_In_ const OrtMemoryInfo* ptr, _Out_ OrtMemoryInfoDeviceType* out);
 
+  /* \brief Update the OrtEnv instance with custom log severity level
+   *
+   * \param[in] log_severity_level The log severity level.
+   * \param[in] ort_env The OrtEnv instance being used
+   *
+   * \since Version 1.13.
+   */
+  ORT_API2_STATUS(UpdateEnvWithCustomLogLevel, OrtLoggingLevel log_severity_level, _In_ const OrtEnv* ort_env);
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:3600]: You can't really have OrtEnv const when you've an API to update it :-).

Comment (yf711) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Thanks for pointing out! That has been fixed

File: csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.shared.cs [hunk 2]
```diff
@@ -252,6 +253,13 @@ public struct OrtApi
         public IntPtr ReleaseKernelInfo;
 
         public IntPtr GetTrainingApi;
+        public IntPtr SessionOptionsAppendExecutionProvider_CANN;
```
Comment (chilo-ms) [inline csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.shared.cs:256]: Are these CANN related declarations expected to be here?

Comment (yf711) [reply csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.shared.cs:256]: Yes, as all newly added C APIs before mine would also need to be added here, so that the same order could be kept.
 
I initially learned from this previous [commit](https://github.com/microsoft/onnxruntime/commit/0dad79b495cd9506fc18c5973114009546c5f9e2) to add additional APIs to keep order same as C APIs, and this was also mentioned in the [code here](https://github.com/microsoft/onnxruntime/blob/234e7551ca3dda73eee074d42bbdf998ecc07481/csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.shared.cs#L16-L17)

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (yuslepukhin) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Please, add C++ API so users do not have to go to C back and forth.

File: csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs [hunk 3]
```diff
@@ -150,6 +151,26 @@ public string[] GetAvailableProviders()
...
+
+
+        /// <summary>
+        /// Get current log level of OrtEnv instance
+        /// </summary>
+        /// <returns>current log level</returns>
+        public LogLevel GetLogLevel()
+        {
+            return currentLogLevel;
+        }
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs:163]: If this is really needed, it should be a property. And you can then combine set/get

**Review by yuslepukhin — CHANGES_REQUESTED** [review]
:clock1:

File: csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs (hunk 3 above)
Comment (yf711) [reply csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs]: Thanks for sharing! This has been improved in the [latest commit ](https://github.com/microsoft/onnxruntime/pull/13402/commits/9715a16b2b77b5a6ff59326c981c433e66b9c0d5)and verified in test case.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (yf711) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: C++ API has been added in [latest commit ](https://github.com/microsoft/onnxruntime/pull/13402/commits/9715a16b2b77b5a6ff59326c981c433e66b9c0d5)as well. Thanks!

Comment (yuslepukhin) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: One important thing to mention is thread-safety. Typically, we do not update the environment after the creation and any thread or even multiple sessions can use it.

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 4]
```diff
@@ -167,6 +167,16 @@ ORT_API_STATUS_IMPL(OrtApis::DisableTelemetryEvents, _In_ const OrtEnv* ort_env)
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(OrtApis::UpdateEnvWithCustomLogLevel, OrtLoggingLevel log_severity_level,
+                    _In_ OrtEnv* ort_env) {
+  API_IMPL_BEGIN
+  LoggingManager* default_logging_manager = ort_env->GetLoggingManager();
+  int severity_level = static_cast<int>(log_severity_level);
+  default_logging_manager->SetDefaultLoggerSeverity(static_cast<logging::Severity>(severity_level));
```
Comment (yuslepukhin) [inline onnxruntime/core/session/onnxruntime_c_api.cc:175]: >default_logging_manager->SetDefaultLoggerSeverity(static_cast<logging::Severity>(severity_level)); [](http://example.com/codeflow?start=2&length=98)

Need to look at the thread-safety here

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (yf711) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: > One important thing to mention is thread-safety. Typically, we do not update the environment after the creation and any thread or even multiple sessions can use it.

Really appreciate the context you shared. The reason adding this env_update feature is this feature exists in Pybind but missing in c# binding, and we would like to unify the user experience and provide the ability to adjust default logger

This feature reuses the existing design in [Pybind:set_default_logger_severity](https://github.com/microsoft/onnxruntime/blame/952c99304a764328070ac6c07f510deec06dedf4/onnxruntime/python/onnxruntime_pybind_state.cc#L923-L924). Do you have any suggestions in order to improve this design?

**Review by yuslepukhin — APPROVED** [review]
:shipit:
